File: solutions/javascript/print-balanced-brace-combinations.js
let printAllBracesRec = function(n, leftCount, rightCount, output, result) {

  if (leftCount >= n && rightCount >= n) {
    result.push(output.slice());
  }

  if (leftCount < n) {
    output.push('{');
    printAllBracesRec(n, leftCount + 1, rightCount, output, result);
    output.pop();
  }
  if (rightCount < leftCount) {
    output.push('}');
    printAllBracesRec(n, leftCount, rightCount + 1, output, result);
    output.pop();
  }
};

let printAllBraces = function(n) {
  let output = [];
  let result = [];
  printAllBracesRec(n, 0, 0, output, result);
  return result;
};

let output = printAllBraces(3);
console.log(output)

File: solutions/python/print-balanced-brace-combinations.py
def print_all_braces_rec(n, left_count, right_count, output, result):
  if left_count >= n and right_count >= n:
    result.append(copy.copy(output));
    
  if left_count < n:
    output += '{'
    print_all_braces_rec(n, left_count + 1, right_count, output, result)
    output.pop()

  if right_count < left_count:
    output += '}'
    print_all_braces_rec(n, left_count, right_count + 1, output, result)
    output.pop()

def print_all_braces(n):
  output = []
  result = []
  print_all_braces_rec(n, 0, 0, output, result)
  return result

result = print_all_braces(3)

for rs in result: 
  print(rs)
